providers/common/kdfs/kbkdf.c

```diff
@@ -0,0 +1,332 @@
+                                unsigned char *context, size_t context_len,
+                                unsigned char *k_i, size_t h, uint32_t l,
+                                unsigned char *ko, size_t ko_len)
+{
+    int ret = 0;
+    EVP_MAC_CTX *ctx = NULL;
+    size_t written = 0, to_write;
+    const unsigned char zero = 0;
+    uint32_t counter, i;
+
+    for (counter = 1; written < ko_len; counter++) {
+        i = be32(counter);
+
+        ctx = EVP_MAC_CTX_dup(ctx_init);
+        if (ctx == NULL)
+            goto done;
+
+        if (!EVP_MAC_update(ctx, (unsigned char *)&i, 4)
+            || !EVP_MAC_update(ctx, label, label_len)
+            || !EVP_MAC_update(ctx, &zero, 1)
+            || !EVP_MAC_update(ctx, context, context_len)
+            || !EVP_MAC_update(ctx, (unsigned char *)&l, 4)
+            || !EVP_MAC_final(ctx, k_i, NULL, h))
+            goto done;
+
+        to_write = ko_len - written;
+        memcpy(ko + written, k_i, MIN(to_write, h));
+        written += h;
+
+        EVP_MAC_CTX_free(ctx);
+        ctx = NULL;
+    }
+
+    ret = 1;
+done:
+    EVP_MAC_CTX_free(ctx);
+    return ret;
+}
+
+static int kbkdf_derive(void *vctx, unsigned char *key, size_t keylen)
+{
+    KBKDF *ctx = (KBKDF *)vctx;
+    int ret = 0;
+    unsigned char *k_i = NULL;
+    uint32_t l = be32(keylen * 8);
+    size_t h = 0;
+
+    /* Label and Context are permitted to be empty. Check everything else. */
+    if (ctx->ctx_init == NULL) {
+        if (ctx->ki_len == 0 || ctx->ki == NULL) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_NO_KEY_SET);
+            return 0;
+        } else if (ctx->mac == NULL) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_MAC);
+            return 0;
+        }
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_MESSAGE_DIGEST);
+        return 0;
+    }
+
+    h = EVP_MAC_size(ctx->ctx_init);
+    if (h == 0)
+        goto done;
+
+    k_i = OPENSSL_zalloc(h);
+    if (k_i == NULL)
+        goto done;
+
+    ret = kbkdf_derive_counter(
+        ctx->ctx_init, ctx->label, ctx->label_len, ctx->context,
+        ctx->context_len, k_i, h, l, key, keylen);
+done:
+    if (ret != 1)
+        OPENSSL_cleanse(key, keylen);
+    OPENSSL_clear_free(k_i, h);
+    return ret;
+}
+
+static int kbkdf_set_buffer(unsigned char **out, size_t *out_len,
+                            const OSSL_PARAM *p)
+{
+    if (p->data == NULL || p->data_size == 0)
+        return 1;
+
+    OPENSSL_clear_free(*out, *out_len);
+    *out = NULL;
+    return OSSL_PARAM_get_octet_string(p, (void **)out, 0, out_len);
+}
+
+static int kbkdf_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    KBKDF *ctx = (KBKDF *)vctx;
+    OPENSSL_CTX *libctx = PROV_LIBRARY_CONTEXT_OF(ctx->provctx);
+    const OSSL_PARAM *p;
+    const EVP_MD *md;
+    EVP_MAC *mac;
+    EVP_MAC_CTX *ctx_init;
+    OSSL_PARAM mparams[3];
+
+    if (!ossl_prov_digest_load_from_params(&ctx->digest, params, libctx))
```

Comment:
If you use `ossl_prov_macctx_load_from_params`, you can throw away `ctx->digest` and `ctx->mac`. Basically, that routine will set up the MAC context for you with the MAC it finds and feed it the appropriate digest setup params.

To then check that you got the right MAC, all you have to do is this:

``` C
    if (EVP_MAC_is_a(EVP_MAC_CTX_mac(ctx->ctx_init), OSSL_MAC_NAME_HMAC)) 
        .... 
```

Reply:
Sorry, still confused - there's no such function `ossl_prov_macctx_load_from_params()`.  I suppose I could implement it - is that what you're suggesting?

Reply:
Ah, I see you added it in #9946.  I'll rebase on top of that.